lib/src/core/engine.dart

```diff
@@ -192,49 +187,61 @@ class Engine extends Disposable with EventsEmittable<EngineEvent> {
   Future<void> sendDataPacket(
     lk_models.DataPacket packet,
   ) async {
-    // make sure we do have a data connection
-    await _ensurePublisherConnected();
+    //
+    rtc.RTCDataChannel? publisherDataChannel(Reliability reliability) =>
+        reliability == Reliability.reliable ? _reliableDCPub : _lossyDCPub;
+
+    rtc.RTCDataChannelState publisherDataChannelState(
+            Reliability reliability) =>
+        publisherDataChannel(reliability)?.state ??
+        rtc.RTCDataChannelState.RTCDataChannelClosed;
+
+    final reliability = packet.kind.toSDKType();
 
     // construct the data channel message
     final message =
         rtc.RTCDataChannelMessage.fromBinary(packet.writeToBuffer());
 
-    // chose data channel
-    final rtc.RTCDataChannel? channel =
-        packet.kind == lk_models.DataPacket_Kind.LOSSY ? _lossyDC : _reliableDC;
+    if (_subscriberPrimary) {
+      // make sure publisher transport is connected
 
-    // send if channel exists
-    if (channel == null) {
-      throw UnexpectedStateException('Data channel is not ready');
-    }
+      if (publisher?.pc.iceConnectionState?.isConnected() != true) {
+        logger.fine('Publisher is not connected...');
```

Comment:
What happens if there are a bunch of data channel messages sent by the app in this state? Are all of them becoming promises and get sent (promises resolved) when the data channel connects? Or are the messages just dropped on the floor?

Reply:
promises should be pending until `waitFor` condition clears (received notification) or timeouts.
actually maybe there should be more checks to prevent `negotiate()` being called multiple times.. 🤔
(not sure how immediate the state changes to `RTCIceConnectionStateChecking`)